src/main/java/net/logstash/logback/composite/FastISOTimestampFormatter.java

```diff
@@ -0,0 +1,359 @@
+         * 
+         * @param tstamp the timestamp millis
+         * @return {@code true} if this state is applicable to the given timestamp
+         */
+        public boolean isApplicable(long tstamp) {
+            return tstamp >= this.zoneTransitionStart && tstamp < this.zoneTransitionStop;
+        }
+        
+        
+        /**
+         * Format the timestamp millis.
+         * Note: you must first invoke {@link #isApplicable(long)} to check that the state can be used to format the timestamp.
+         * 
+         * @param tstamp the timestamp milis to format
+         * @return the formatted timestamp
+         */
+        public String format(long tstamp) {
+            // If caching is disabled...
+            //
+            if (!this.cachingEnabled) {
+                return buildFromFormatter(tstamp);
+            }
+            
+            // If tstamp is within the caching period...
+            //
+            TimestampCache currentCache = this.cache;
+            if (currentCache != null && currentCache.isApplicable(tstamp)) {
+                return buildFromCache(currentCache, tstamp);
+            }
+            
+            // ... otherwise, use the formatter and cache the formatted value
+            //
+            String formatted = buildFromFormatter(tstamp);
+            cache = createNewCache(tstamp, formatted);
+            return formatted;
+        }
+        
+        
+        private TimestampCache createNewCache(long timestampInMillis, String formatted) {
+            // The part up to the minutes (included)
+            String prefix = formatted.substring(0, 17);
+                
+            // The part after the millis (i.e. the timezone)
+            // Note:
+            //   ISO_LOCAL_DATE_TIME has nothing after millis
+            //   ISO_INSTANT has a 'Z'
+            int pos = formatted.indexOf('+', 17);
+            if (pos == -1) {
+                pos = formatted.indexOf('-', 17);
+            }
+            if (pos == -1 && formatted.charAt(formatted.length() - 1) == 'Z') {
+                pos = formatted.length() - 1;
+            }
+            String suffix = pos == -1 ? "" : formatted.substring(pos);
+            
+            // Determine how long we can use this cache
+            long cachePeriod = timestampInMillis / 60_000;
+            long cacheStart = cachePeriod * 60_000;
+            long cacheStop = (cachePeriod + 1) * 60_000;
+
+            // Store in cache
+            return new TimestampCache(cacheStart, cacheStop, prefix, suffix);
+        }
+        
+        
+        private String buildFromCache(TimestampCache cache, long timestampInMillis) {
+            return cache.format(timestampInMillis);
+        }
+    }
+    
+    String buildFromFormatter(long timestampInMillis) {
+        return FastISOTimestampFormatter.this.formatter.format(Instant.ofEpochMilli(timestampInMillis));
+    }
+    
+    private class TimestampCache {
```

Comment:
Hmm.  Naming...  I tend to think of a Cache as a collection of "things", but this is just one "thing".

Perhaps `TimestampPeriod` rather than `TimestampCache` ?

And `ZoneOffsetState.cache` could be `ZoneOffsetState.cachedTimestampPeriod`